runtime/domain.c

```diff
@@ -1111,29 +1151,34 @@ static void* domain_thread_func(void* v)
     caml_gc_log("Domain starting (unique_id = %"ARCH_INTNAT_PRINTF_FORMAT"u)",
                 domain_self->interruptor.unique_id);
     CAML_EV_LIFECYCLE(EV_DOMAIN_SPAWN, getpid());
-    /* FIXME: ignoring errors and asynchronous exceptions during
-       domain initialization is unsafe and/or can deadlock. */
+    /* FIXME: ignoring errors during domain initialization is unsafe
+       and/or can deadlock. */
     caml_domain_initialize_hook();
 
     /* release callback early;
        see the [note about callbacks and GC] in callback.c */
     value unrooted_callback = ml_values->callback;
     caml_modify_generational_global_root(&ml_values->callback, Val_unit);
-    caml_callback_exn(unrooted_callback, Val_unit);
+    value res_or_exn = caml_callback_exn(unrooted_callback, Val_unit);
+    domain_sync_result(ml_values->term_sync, res_or_exn);
```

Comment:
Now that the handling of `res_or_exn` is more regular, it would be easy to convert `res_or_exn` into an OCaml value here, caller-side, instead of doing the work in `domain_sync_result`. (This also avoids the awkwardness of passing `domain_sync_result` a `value` parameter that cannot validly be rooted.)

Reply:
No thanks, this makes `domain_thread_func` even bigger (and if you introduce a helper function you are back to the awkardness of passing an irregular value).

Reply:
```c
value result = caml_extract_result(caml_callback_exn(unrooted_callback, Val_unit));
```

> if you introduce a helper function you are back to the awkardness of passing an irregular value

It is easier to be careful about a short helper function that is specifically dedicated to a subtle input type than about a larger function that takes one weird arguments among other things.